### PR TITLE
[WIP] Add support for transpile packages installed via NPM 

### LINKED ISSUE #706: Add support to babel transpile modules inside node_modules
Now some of us ships NPM packages (specially components) written in ES2015 without transpiling them.

That's a pretty good thing specially if they are gonna used in a project like Next.js or CRA (which does transpiling). They offer benefits like:

* No need to transpile before shipping to NPM
* Get the benefit of Webpack 2's tree-shaking

But we can't do this now we [exclude](https://github.com/zeit/next.js/blob/master/server/build/webpack.js#L166) everything inside node_modules from babel transpiling.

So, here's the proposed solution. 

We have an entry in `next.config.js` to include modules which needs to go through babel like this:

```js
module.exports = {
   babelifyModuleRoots: [
     "my-component",
     "redux/src"
   ]
}
```

### PR DESCRIPTION
**Pull request by arunoda**
Fixes #706

**Instructions**

Add a field called `transpileModules` with an array of regexp patterns like this:

```js
module.exports = {
 transpileModules : [
    /react-button/
  ]
}
```

Here Next.js checks every imported module with above patterns and transpile them if matched. By default it uses `next/babel` preset, but you can use a `.babelrc` file to customize it as you want.

In the above example, Next.js will transpile any module inside the `node_modules` directory which has "react-button" in it's pathname.

> Note: Currently, Next.js only supports modules installed into `node_modules`. So, transpiling modules installed with `npm link <my-comp-name>` doesn't work well.

### CODE REVIEW

File: server/index.js [hunk 1]
```diff
@@ -30,6 +31,13 @@ export default class Server {
     this.config = getConfig(this.dir)
 
     this.defineRoutes()
+    babelRegister({
+      // If there's .babelrc file in the app path(or similar) it'll take the
+      // precedence over this.
+      presets: [require.resolve('./build/babel/preset')],
+      extensions: ['.js'],
+      only: this.config.transpileModules || []
+    })
```
Comment (nkzawa) [inline server/index.js]: Umm, babel-register is not for production. I'm not sure if it's acceptable.

Comment (arunoda) [inline server/index.js]: I am not sure about that too. Need to read the source to dig more.

Comment (arunoda) [inline server/index.js]: Yep. I think we can go some kind of pre-compilation system. 
For now, we can use this to test the functionality.

File: README.md [hunk 2]
```diff
@@ -522,6 +522,36 @@ Here's an example `.babelrc` file:
...
+    /react-button/
+  ]
+}
+```
+
+Here Next.js checks every imported module with above patterns and transpile them if matched. By default it uses `next/babel` preset, but you can use a `.babelrc` file to customize it as you want.
+
+In the above example, Next.js will transpile any module inside the `node_modules` directory which has "react-button" in it's pathname.
+
+> Note: Currently, Next.js only supports modules installed into `node_modules`. So, transpiling modules installed with `npm link <my-comp-name>` doesn't work well.
```
Comment (slorber) [inline README.md]: so it does not work with symlinks (like Lerna do?) not sure to understand that comment

Comment (arunoda) [inline README.md]: For now yes it is.

**Comment by abstracthat** [comment]
Where are we at with this feature? I just ran into this problem and would be lovely to not have to add transpiled code to the component repo.

**Comment by ajoslin** [comment]
I need this to work with symlink support (my project makes heavy use of Lerna).

I'd love to help contribute to get symlink support up and running. What steps need to be taken?

**Comment by traa** [comment]
Any possible workarounds to use Next.js with Lerna?

**Comment by timneutkens** [comment]
@arunoda do we still want to take this?

**Comment by arunoda** [comment]
Let's keep this open. I want to work on this someday.

**Comment by esteban-uo** [comment]
an status on this issue @arunoda ? thx!

File: server/index.js (hunk 1 above)
Comment (jribeiro) [inline server/index.js]: https://github.com/standard-things/esm looks promising. Maybe we could use this in production and `babel-register` in dev mode?

File: server/build/webpack.js [hunk 3]
```diff
@@ -156,6 +156,16 @@ export default async function createCompiler (dir, { dev = false, quiet = false
...
+        if (!(pattern instanceof RegExp)) {
+          const message = `Incorrect pattern in config.transpileModules: "${pattern}".` +
+            'It accepts an array of regular expression'
+          throw new Error(message)
+        }
+
+        if (pattern.test(str)) return false
+      }
+
       return /node_modules/.test(str)
```
Comment (jribeiro) [inline server/build/webpack.js]: would be cool if this was more generic to support monorepo scenarios. I'm happy to lend a hand to this feature if needed

**Comment by thealjey** [comment]
will this ever get merged?

**Comment by mattfelten** [comment]
What happened here? Looks like the feature died?

**Comment by rickbyington** [comment]
Yes, would really like my Lerna packages to be able to get transpiled.

**Comment by philcockfield** [comment]
@timneutkens when you closed this, was it merged?

**Comment by timneutkens** [comment]
@philcockfield no, check https://github.com/zeit/next.js/pull/3319

**Comment by philcockfield** [comment]
Thanks @timneutkens

**Pull request closed without merging**